### ISSUE TITLE
KDF-bound in, KDF-bound out

### ISSUE THREAD

**Issue by martinthomson** [tlswg/draft-ietf-tls-external-psk-importer#16]
In reviewing the draft, it was not clear that the input PSK was also bound to a hash function, but that the process could produce a derived PSK bound to any hash function.  As this is the primary function of the document, that seems problematic.

This seems to be the answer to this:

> In particular, it describes how KDF-bound PSKs can be differentiated by different hash algorithms to produce a set of candidate PSKs, each of which are bound to a specific hash function.

I assume that this means KDF rather than hash function throughout (the KDF-bound thing seems right).  But it never really comes out and says that:

> This process takes a PSK, which is bound to a specific KDF.  The output of this process is a PSK that is bound to a TLS KDF.  This document only describes output KDF functions that are defined in TLS, so the TLS 1.2 PRF or HKDF in TLS 1.3."

**Comment by chris-wood** [comment]
> it was not clear that the input PSK was also bound to a hash function

It's optionally bound to a hash function. In defining external PSKs, the document states that the "associated KDF (and hash function) may be undefined." Do you think it would be more clear to put this in the introduction? (Minimally, as you point out, the introduction should not assume that the KDF-bound PSK as input is not quite right, so that should probably change.)

As for the second comment (KDF output), that's not quite the goal. The importer binds input PSKs to a hash via HKDF, not via the TLS KDF. (The TLS 1.2 PRF is not used.) 

To summarize, my view is that the contract is more or less: optional hash-bound PSK in, hash-bound PSK out. Does that seem incorrect to you?

**Comment by martinthomson** [comment]
Hmm, my understanding is that we don't have any support for optional bindings on input.  I'm inclined to say that we need stricter constraints on input.  Namely that they are bound to a specific KDF and that KDF needs to take a context struct in the form this document describes.

Yes, limiting the concrete instantiation of this to HKDF is fine (and the only sensible option).  However, if you don't bind the input key more tightly, you will end up shifting the problem you set out to solve one step further out.

**Comment by chris-wood** [comment]
> Namely that they are bound to a specific KDF and that KDF needs to take a context struct in the form this document describes.

I'm not sure I follow anymore! In my view, the problem we aim to solve is one where the external PSK was *not* provisioned "correctly," i.e., with the right hash (KDF) function in mind. It's just a plain old n-bit key. The importer is designed to bind this key to a specific hash (KDF). 

If the machinery that produced external PSKs was forced to change to accommodate a KDF, then it seems that machinery could also vend PSKs that do precisely what the importer does. Namely, differentiate by (context, protocol, hash/KDF). 

Perhaps I'm misunderstanding your point. Can you please clarify? If you're looking for a specific change, could you please make a suggestion?

**Comment by russhousley** [comment]
In TLS 1.2 and earlier, PSKs were not bound to a KDF.  Now in TLS 1.3, a PSK is required to be bound to a KDF.  I thought the goal was to start with a PSK that is not bound to a KDF and produce a PSK that can be used in the desired context.  So, the input is always a "plain old n-bit key" (Chris' words).  The output, when the desired context is TLS 1.3, must include the KDF.

**Comment by chris-wood** [comment]
> The output, when the desired context is TLS 1.3, must include the KDF.

This is *exactly* what the importer does. Do you disagree?

**Comment by russhousley** [comment]
It is not clear to me what the importer does if I want the context to be TLS 1.2.

**Comment by chris-wood** [comment]
It changes the info (or proposed `protocol` field) value and runs the key through a KDF.

**Comment by russhousley** [comment]
Right, I get that, which I think will bind the resulting key to TLS 1.2 but allow any KDF.  I think the document should be explicit if that is the case.

**Comment by martinthomson** [comment]
Looking at #17, I think that you want to bind this to the TLS 1.2 KDF differently by using a different KDF identifier.  The identifiers you have are only good for TLS 1.3.

You probably also want security considerations addressing the fact that the TLS 1.2 has this odd characteristic.

**Comment by chris-wood** [comment]
>  I think that you want to bind this to the TLS 1.2 KDF...

Can you please elaborate on why you think this is necessary? 

As an example, let's assume I have an external PSK and I want to import it for TLS 1.2 and 1.3, and I only want to support one hash function (or KDF). Let's say that KDF is HKDF_SHA256. I would then import with the tuples (0x0303, HKDF_SHA256) and (0x0304, HKDF_SHA256), and it seems I'd have the desired separation.

In other words, I don't see a technical reason why importing PSKs for TLS 1.2 requires use of the TLS 1.2 KDF (PRF).

**Comment by russhousley** [comment]
I think we are saying the same thing, except that when the protocol version is TLS 1.2 that it will be used with the PRF that is negotiated in the TLS 1.2 handshake.  I just think the document should say that explicitly.

**Comment by chris-wood** [comment]
> except that when the protocol version is TLS 1.2 that it will be used with the PRF that is negotiated in the TLS 1.2 handshake

This is *not* what I'm saying! I'm saying *do not use the TLS 1.2 PRF*. Rather, use HKDF. Can you clarify why you think the TLS 1.2 PRF should be used instead?

**Comment by russhousley** [comment]
There are two thongs that are being confused here.  First, the one used to create the protocol-version-specific key.  This is the one that you want to always use HKDF.  That is fine with me.  Second, with TLS 1.3, it is bound to a particular hash function.  I am saying that when a key is derived for TLS 1.2 the document should say that it will be used with whatever PRF is negotiated in the TLS handshake.  Right now, I do d not see anything about that in the document at all.

**Comment by chris-wood** [comment]
Thanks for clarifying! 

> I am saying that when a key is derived for TLS 1.2 the document should say that it will be used with whatever PRF is negotiated in the TLS handshake.

If I use an external PSK for TLS 1.2 today, without any import step, it's already the case that it's used with whatever PRF is negotiated in the handshake. The importer does not change that, and therefore I think doesn't need to say anything about that.

**Comment by russhousley** [comment]
I just think it would be better say it since TLS 1.3 had a binding but TLS 1.2 does not.  You now understand my point, so I will leave it to you as author.

**Comment by chris-wood** [comment]
Closed via #18.
